### DISCUSSION TITLE
A New Date Picker Control

### DISCUSSION THREAD

**Discussion by jrovny** [CommunityToolkit/Labs-Windows#289] [Ideas]
The existing Microsoft.UI.Xaml.Controls Date Picker as documented [here](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.datepicker?view=winrt-22621) is unworkable for us as it doesn't allow for rapid data entry. The user is _forced_ to interact with the control using the mouse. That alone is a non-starter for our users, not to mention the control has a very unintuitive look and feel.

![XamlDatePickerCrossedOut2](https://user-images.githubusercontent.com/13325610/196790298-e42ca7a6-5910-4ff7-88e0-eee166ffbb19.png)

It would be extremely helpful if there was a date picker that allows data entry much like a regular textbox, but with a mini calendar button embedded in it that opens a calendar control. As far as behavior, the Angular Material date picker is a good reference point. It auto-formats the dates using a default date format template. You can even specify a collection of date format templates to use. So if a user entered the date as "dd/MM," "yyyy-MM-dd" or "yy-MM-dd" the control will understand and format accordingly. It also flags invalid dates.

This is a control I plan to create for my organization. I wondered if something like this was ever considered or potentially in the works.  
I searched but couldn't find anything. 

Very interested in your thoughts and feedback.

Thanks,

**Comment by michael-hawker** [comment]
Thanks for the proposal @jrovny. The closest one would be this issue on the platform repo: https://github.com/microsoft/microsoft-ui-xaml/issues/735 for `CalendarDatePicker` to have an `IsEditable` property to allow direct text entry.

I am able to use the keyboard with `DatePicker` and `CalendarDatePicker`, so I think your concern is just being around being able to type a date directly (which is quicker and easier for some folks)?

Luckily, `CalendarDatePicker` isn't sealed, so I would suggest trying:
- subclassing directly from it
- add new `IsEditable` dependency property
- [customize the template](https://github.com/microsoft/microsoft-ui-xaml/blob/main/dev/CommonStyles/CalendarDatePicker_themeresources.xaml)
  - Make it contain a textbox for date entry (could bring over the [toolkit regex helpers maybe](https://learn.microsoft.com/en-us/windows/communitytoolkit/extensions/textboxextensions)?)
  - Make the calendar icon a button which brings up the existing UI?
  - Hook events/things up in the new class to make work as desired

Not sure what behavior is baked into the existing control with click events and stuff, but this would be the best way to create this thing and make it easier to maintain without having to duplicate all the existing behavior. (I'm actually trying to do something similar for the `Expander` control to support our Left/Right directions but on top of the WinUI version of the control.)

If you'd like to try doing this in labs, let us know. We're here to help.
